File: packages/react-native/android/src/main/java/com/rejourney/recording/ViewHierarchyScanner.kt
/**
 * Copyright 2026 Rejourney
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.rejourney.recording

import android.app.Activity
import android.content.Context
import android.graphics.Rect
import android.os.SystemClock
import android.view.View
import android.view.ViewGroup
import android.widget.*
import java.lang.ref.WeakReference

/**
 * View hierarchy scanning and serialization
 * Android implementation aligned with iOS ViewHierarchyScanner.swift
 */
class ViewHierarchyScanner private constructor() {
    
    companion object {
        @Volatile
        private var instance: ViewHierarchyScanner? = null
        
        val shared: ViewHierarchyScanner
            get() = instance ?: synchronized(this) {
                instance ?: ViewHierarchyScanner().also { instance = it }
            }
    }
    
    var maxDepth: Int = 12
    var includeTextContent: Boolean = true
    var includeVisualProperties: Boolean = true
    
    private val timeBudgetMs: Long = 16 // 16ms to stay under one frame
    
    private var currentActivity: WeakReference<Activity>? = null
    
    fun setCurrentActivity(activity: Activity?) {
        currentActivity = if (activity != null) WeakReference(activity) else null
    }
    
    fun captureHierarchy(): Map<String, Any>? {
        val activity = currentActivity?.get() ?: return null
        val decorView = activity.window?.decorView ?: return null
        return serializeWindow(decorView, activity)
    }
    
    fun serializeWindow(window: View, activity: Activity): Map<String, Any> {
        val ts = System.currentTimeMillis()
        val displayMetrics = activity.resources.displayMetrics
        val bounds = Rect().also { window.getWindowVisibleDisplayFrame(it) }
        val startTime = SystemClock.elapsedRealtime()
        
        val root = serializeView(window, 0, startTime) ?: emptyMap()
        
        val result = mutableMapOf<String, Any>(
            "timestamp" to ts,
            "screen" to mapOf(
                "width" to bounds.width(),
                "height" to bounds.height(),
                "scale" to displayMetrics.density
            ),
            "root" to root
        )
        
        ReplayOrchestrator.shared?.currentScreenName?.let {
            result["screenName"] = it
        }
        
        return result
    }
    
    private fun serializeView(view: View, depth: Int, startTime: Long): Map<String, Any>? {
        if (depth > maxDepth) return null
        if (SystemClock.elapsedRealtime() - startTime > timeBudgetMs) {
            return mapOf("type" to view.javaClass.simpleName, "bailout" to true)
        }
        if (depth > 0 && (!view.isShown || view.alpha <= 0.01f || view.width <= 0 || view.height <= 0)) {
            return null
        }
        
        val node = mutableMapOf<String, Any>()
        node["type"] = view.javaClass.simpleName
        
        val location = IntArray(2)
        view.getLocationInWindow(location)
        node["frame"] = mapOf(
            "x" to location[0],
            "y" to location[1],
            "w" to view.width,
            "h" to view.height
        )
        
        if (!view.isShown) node["hidden"] = true
        if (view.alpha < 1.0f) node["alpha"] = view.alpha
        
        // Get accessibility identifier / test ID
        view.contentDescription?.toString()?.takeIf { it.isNotEmpty() }?.let {
            node["testID"] = it
        }
        
        // Check for React Native nativeID
        try {
            val nativeId = view.getTag(com.facebook.react.R.id.view_tag_native_id) as? String
            if (!nativeId.isNullOrEmpty()) {
                node["testID"] = nativeId
            }
        } catch (_: Exception) { }
        
        if (isSensitive(view)) node["masked"] = true
        
        if (includeVisualProperties) {
            view.background?.let { bg ->
                // Try to get background color
                try {
                    val colorDrawable = bg as? android.graphics.drawable.ColorDrawable
                    colorDrawable?.color?.let { color ->
                        node["bg"] = String.format("#%06X", 0xFFFFFF and color)
                    }
                } catch (_: Exception) { }
            }
        }
        
        if (includeTextContent) {
            when (view) {
                is TextView -> {
                    val text = view.text?.toString() ?: ""
                    node["text"] = maskText(text)
                    node["textLength"] = text.length
                    
                    if (view is EditText) {
                        node["text"] = "***"
                        view.hint?.toString()?.let { node["placeholder"] = it }
                    }
                }
            }
        }
        
        if (isInteractive(view)) {
            node["interactive"] = true
            
            when (view) {
                is Button -> {
                    node["buttonTitle"] = view.text?.toString() ?: ""
                    node["enabled"] = view.isEnabled
                }
                is CompoundButton -> {
                    node["checked"] = view.isChecked
                    node["enabled"] = view.isEnabled
                }
            }
            
            if (view.isEnabled) {
                node["enabled"] = true
            } else {
                node["enabled"] = false
            }
        }
        
        if (view is ScrollView || view is HorizontalScrollView) {
            node["scrollEnabled"] = true
            node["contentOffset"] = mapOf<String, Any>(
                "x" to ((view as? HorizontalScrollView)?.scrollX ?: (view as? ScrollView)?.scrollX ?: 0),
                "y" to ((view as? HorizontalScrollView)?.scrollY ?: (view as? ScrollView)?.scrollY ?: 0)
            )
        }
        
        if (view is ImageView) {
            node["hasImage"] = view.drawable != null
        }
        
        // Process children
        if (view is ViewGroup) {
            val children = mutableListOf<Map<String, Any>>()
            for (i in 0 until view.childCount) {
                val child = view.getChildAt(i)
                if (child.isShown && child.alpha > 0.01f) {
                    serializeView(child, depth + 1, startTime)?.let {
                        children.add(it)
                    }
                }
            }
            if (children.isNotEmpty()) {
                node["children"] = children
            }
        }
        
        return node
    }
    
    private fun isSensitive(view: View): Boolean {
        if (view is EditText) {
            val inputType = view.inputType
            // Check for password input types
            if (inputType and android.text.InputType.TYPE_TEXT_VARIATION_PASSWORD != 0 ||
                inputType and android.text.InputType.TYPE_TEXT_VARIATION_VISIBLE_PASSWORD != 0 ||
                inputType and android.text.InputType.TYPE_NUMBER_VARIATION_PASSWORD != 0) {
                return true
            }
        }
        return false
    }
    
    private fun isInteractive(view: View): Boolean {
        return view is Button ||
               view is EditText ||
               view is CheckBox ||
               view is RadioButton ||
               view is Switch ||
               view is SeekBar ||
               view is Spinner ||
               view.isClickable
    }
    
    private fun maskText(text: String): String {
        return if (text.length > 100) text.take(100) + "..." else text
    }
}

File: packages/react-native/ios/Recording/ViewHierarchyScanner.swift
/**
 * Copyright 2026 Rejourney
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import UIKit

@objc public class ViewHierarchyScanner: NSObject {
    
    @objc public static let shared = ViewHierarchyScanner()
    @objc public var maxDepth: Int = 12
    @objc public var includeTextContent: Bool = true
    @objc public var includeVisualProperties: Bool = true
    
    private let _timeBudgetNs: UInt64 = 16_000_000
    
    private override init() {
        super.init()
    }
    
    @objc public func captureHierarchy() -> [String: Any]? {
        guard let w = _keyWindow() else { return nil }
        // Only serialize the key window — skip keyboard/system windows
        // whose internal views produce NaN frames during transition
        return serializeWindow(w)
    }
    
    @objc public func serializeWindow(_ window: UIWindow) -> [String: Any] {
        let ts = Int64(Date().timeIntervalSince1970 * 1000)
        let scale = window.screen.scale
        let bounds = window.bounds
        let start = DispatchTime.now().uptimeNanoseconds
        let root = _serializeView(window, depth: 0, start: start)
        
        var result: [String: Any] = [
            "timestamp": ts,
            "screen": ["width": bounds.width, "height": bounds.height, "scale": scale],
            "root": root ?? [:]
        ]
        
        if let sn = ReplayOrchestrator.shared.currentScreenName {
            result["screenName"] = sn
        }
        return result
    }
    
    private func _keyWindow() -> UIWindow? {
        if #available(iOS 15.0, *) {
            return UIApplication.shared.connectedScenes.compactMap { $0 as? UIWindowScene }.flatMap { $0.windows }.first { $0.isKeyWindow }
        } else {
            return UIApplication.shared.windows.first { $0.isKeyWindow }
        }
    }
    
    private func _serializeView(_ view: UIView, depth: Int, start: UInt64) -> [String: Any]? {
        if depth > maxDepth { return nil }
        if (DispatchTime.now().uptimeNanoseconds - start) > _timeBudgetNs { return ["type": _typeName(view), "bailout": true] }
        if depth > 0 && (view.isHidden || view.alpha <= 0.01 || view.bounds.width <= 0 || view.bounds.height <= 0) { return nil }
        
        // Skip keyboard/system windows to avoid NaN frames during keyboard transitions
        let className = _typeName(view)
        if className.contains("UIRemoteKeyboardWindow") ||
           className.contains("UITextEffectsWindow") ||
           className.contains("UIInputSetHostView") ||
           className.contains("UIKeyboard") {
            return nil
        }
        
        var node: [String: Any] = [:]
        node["type"] = className
        
        // Guard frame values against NaN / Inf — keyboard and animated views
        // can have degenerate frames that poison downstream JSON serialization
        let f = view.frame
        let fx = f.origin.x.isFinite && !f.origin.x.isNaN ? f.origin.x : 0
        let fy = f.origin.y.isFinite && !f.origin.y.isNaN ? f.origin.y : 0
        let fw = f.width.isFinite && !f.width.isNaN ? f.width : 0
        let fh = f.height.isFinite && !f.height.isNaN ? f.height : 0
        node["frame"] = ["x": fx, "y": fy, "w": fw, "h": fh]
        
        if view.isHidden { node["hidden"] = true }
        if view.alpha < 1.0 { node["alpha"] = view.alpha }
        if let aid = view.accessibilityIdentifier, !aid.isEmpty { node["testID"] = aid }
        if let lbl = view.accessibilityLabel, !lbl.isEmpty { node["label"] = lbl }
        if _isSensitive(view) { node["masked"] = true }
        
        if includeVisualProperties, let bg = view.backgroundColor, bg != .clear { node["bg"] = _hexColor(bg) }
        
        if includeTextContent {
            if let tv = view as? UITextView { node["text"] = _mask(tv.text ?? ""); node["textLength"] = tv.text?.count ?? 0 }
            else if let lb = view as? UILabel { node["text"] = _mask(lb.text ?? ""); node["textLength"] = lb.text?.count ?? 0 }
            else if let tf = view as? UITextField { node["text"] = "***"; node["textLength"] = tf.text?.count ?? 0; node["placeholder"] = tf.placeholder }
        }
        
        if _isInteractive(view) {
            node["interactive"] = true
            if let btn = view as? UIButton { if let t = btn.title(for: .normal) { node["buttonTitle"] = t }; node["enabled"] = btn.isEnabled }
            if let ctrl = view as? UIControl { node["enabled"] = ctrl.isEnabled }
        }
        
        if let sv = view as? UIScrollView {
            node["scrollEnabled"] = sv.isScrollEnabled
            let ox = sv.contentOffset.x.isFinite && !sv.contentOffset.x.isNaN ? sv.contentOffset.x : 0
            let oy = sv.contentOffset.y.isFinite && !sv.contentOffset.y.isNaN ? sv.contentOffset.y : 0
            let cw = sv.contentSize.width.isFinite && !sv.contentSize.width.isNaN ? sv.contentSize.width : 0
            let ch = sv.contentSize.height.isFinite && !sv.contentSize.height.isNaN ? sv.contentSize.height : 0
            node["contentOffset"] = ["x": ox, "y": oy]
            node["contentSize"] = ["w": cw, "h": ch]
        }
        
        if view is UIImageView { node["hasImage"] = true }
        
        let subs = view.subviews.filter { !$0.isHidden && $0.alpha > 0.01 }
        if !subs.isEmpty {
            var children: [[String: Any]] = []
            for s in subs {
                if let cn = _serializeView(s, depth: depth + 1, start: start) { children.append(cn) }
            }
            if !children.isEmpty { node["children"] = children }
        }
        
        return node
    }
    
    private func _typeName(_ v: UIView) -> String { String(describing: type(of: v)) }
    
    private func _isSensitive(_ v: UIView) -> Bool {
        if v.accessibilityHint == "rejourney_occlude" { return true }
        if let tf = v as? UITextField, tf.isSecureTextEntry { return true }
        return false
    }
    
    private func _isInteractive(_ v: UIView) -> Bool {
        v is UIButton || v is UIControl || v is UITextField || v is UITextView
    }
    
    private func _mask(_ text: String) -> String {
        text.count > 100 ? String(text.prefix(100)) + "..." : text
    }
    
    private func _hexColor(_ c: UIColor) -> String {
        var r: CGFloat = 0, g: CGFloat = 0, b: CGFloat = 0, a: CGFloat = 0
        c.getRed(&r, green: &g, blue: &b, alpha: &a)
        return String(format: "#%02X%02X%02X", Int(r * 255), Int(g * 255), Int(b * 255))
    }
}
